### PR TITLE
chore(promProbe): Adding the prometheus probe

### PR DESCRIPTION
**Pull request by ispeakc0de**
Adding a new probe defined as `PromProbe`, which can be used to run the Prometheus queries and validate the Prometheus metrics output to follow the given criteria.

### sample promeProbe schema

```yaml
probe:
- name: prom-query
  type: "PromProbe"
  promProbe/inputs:
    endpoint: "http://34.66.53.203:31090"
    query: "avg_over_time(probe_success{instance='helloservice.default.svc.cluster.local:80',job='blackbox'}[10s])*100"
    comparator:
      type: "int"
      criteria: "=="
      value: "100"
  mode: "Edge"
  runProperties:
    probeTimeout: 10
    interval: 5
    retry: 1
    probePollingInterval: 2
```

### CODE REVIEW

File: pkg/probe/probe.go [hunk 1]
```diff
@@ -134,6 +141,7 @@ func InitializeProbesInChaosResultDetails(chaosDetails *types.ChaosDetails, clie
 
 	chaosresult.ProbeDetails = probeDetails
 	chaosresult.ProbeArtifacts = map[string]types.ProbeArtifact{}
+	chaosresult.PromProbeImage = Getenv("LIB_IMAGE", "litmuschaos/go-runner:ci")
```
Comment (ksatchit) [inline pkg/probe/probe.go:144]: can change it to latest once we have a "release" with this code.

File: pkg/probe/promeProbe.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+	"github.com/litmuschaos/litmus-go/pkg/types"
+	litmusexec "github.com/litmuschaos/litmus-go/pkg/utils/exec"
+	"github.com/litmuschaos/litmus-go/pkg/utils/retry"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// PreparePromeProbe contains the steps to prepare the prometheus probe
+// which compares the metrices output exposed at the given endpoint
+func PreparePromeProbe(probe v1alpha1.ProbeAttributes, clients clients.ClientSets, chaosDetails *types.ChaosDetails, resultDetails *types.ResultDetails, phase string, eventsDetails *types.EventDetails) error {
```
Comment (ksatchit) [inline pkg/probe/promeProbe.go:21]: We can use same ref as in the schema: promProbe

File: pkg/probe/promeProbe.go [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+			time.Sleep(time.Duration(probe.RunProperties.ProbePollingInterval) * time.Second)
+		}
+	}
+
+}
+
+// ExtractValueFromMetrics extract the value field from the prometheus metrix
+func ExtractValueFromMetrics(metrics string) (string, error) {
+
+	rows := strings.Split(metrics, "\n")
```
Comment (ksatchit) [inline pkg/probe/promeProbe.go]: Lets add some comments here stating explicitly that we are using newline char to differentiate metrics entries

File: pkg/probe/promeProbe.go [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+	if len(rows) > 1 {
+		return "", errors.Errorf("metrics values are greater than one")
+	}
+
+	value := strings.Split(rows[0], ",")
+	if value[2] == "" {
+		return "", errors.Errorf("error while parsing value from derived matrics")
+	}
+
+	return value[2], nil
```
Comment (ksatchit) [inline pkg/probe/promeProbe.go]: does it make sense to typecast to float by default.. and not leave comparator choice to user: https://prometheus.io/docs/introduction/faq/#why-are-all-sample-values-64-bit-floats-i-want-integers

File: pkg/probe/promProbe.go [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+	// output should contains exact one metrics entry along with header
+	// erroring out the cases where it contains more or less entries
+	if len(rows) > 2 {
+		return 0, errors.Errorf("metrics entries can't be more than two")
+	} else if len(rows) < 2 {
+		return 0, errors.Errorf("metrics doesn't contains required values")
+	}
+
+	// deriving the index for the value column from the headers
+	headers := strings.Split(rows[0], ",")
```
Comment (ksatchit) [inline pkg/probe/promProbe.go:331]:
```suggestion
	headerColumn := strings.Split(rows[0], ",")
```

**Review by ksatchit — APPROVED** [review]

**Review by uditgaurav — APPROVED** [review]

**Review by ksatchit — APPROVED** [review]

**Pull request merged**